File: sdks/ruby/lib/cyberarmor/policy.rb
# frozen_string_literal: true

module CyberArmor
  # Policy module provides decision types, the Decision value object, and
  # the error raised when a request is denied in enforce mode.
  module Policy
    # All possible decision types returned by the CyberArmor policy engine.
    DECISION_TYPES = %w[
      ALLOW
      DENY
      ALLOW_WITH_REDACTION
      ALLOW_WITH_LIMITS
      REQUIRE_APPROVAL
      ALLOW_WITH_AUDIT_ONLY
      QUARANTINE
    ].freeze

    # Raised when the policy engine denies a request and the SDK is operating
    # in "enforce" mode.
    class PolicyViolationError < StandardError
      attr_reader :decision_type, :reason

      # @param decision_type [String] one of DECISION_TYPES
      # @param reason [String] human-readable explanation from the policy engine
      def initialize(decision_type:, reason:)
        @decision_type = decision_type
        @reason        = reason
        super("CyberArmor policy violation [#{decision_type}]: #{reason}")
      end
    end

    # Immutable value object representing a policy engine decision.
    Decision = Struct.new(
      :allowed,          # [Boolean] true when the request may proceed
      :decision_type,    # [String]  one of DECISION_TYPES
      :reason,           # [String]  human-readable explanation
      :redacted_prompt,  # [String, nil] prompt after PII/secrets are removed
      keyword_init: true
    ) do
      # Convenience predicate.
      # @return [Boolean]
      def allowed?
        allowed == true
      end

      # Build a Decision from a parsed JSON hash returned by the policy API.
      #
      # @param hash [Hash]
      # @return [Decision]
      def self.from_hash(hash)
        new(
          allowed:         hash.fetch('allowed', false),
          decision_type:   hash.fetch('decision_type', 'DENY'),
          reason:          hash.fetch('reason', ''),
          redacted_prompt: hash['redacted_prompt']
        )
      end
    end
  end
end

File: sdks/rust/src/policy.rs
// cyberarmor-sdk/src/policy.rs
//
// Policy decision types returned by the CyberArmor policy engine.

use serde::{Deserialize, Serialize};

/// All possible decision types the policy engine may return.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum DecisionType {
    /// The request is permitted without restrictions.
    Allow,
    /// The request is denied.
    Deny,
    /// The request is permitted after PII/secrets have been removed from the prompt.
    AllowWithRedaction,
    /// The request is permitted subject to rate or token limits.
    AllowWithLimits,
    /// The request must be routed for human approval before proceeding.
    RequireApproval,
    /// The request is permitted but is recorded for compliance purposes.
    AllowWithAuditOnly,
    /// The request is isolated for investigation.
    Quarantine,
}

impl DecisionType {
    /// Returns `true` if this decision type implies the request should proceed
    /// without raising a policy violation error.
    pub fn is_allowed_variant(&self) -> bool {
        matches!(
            self,
            Self::Allow
                | Self::AllowWithRedaction
                | Self::AllowWithLimits
                | Self::AllowWithAuditOnly
        )
    }
}

impl std::fmt::Display for DecisionType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            Self::Allow               => "ALLOW",
            Self::Deny                => "DENY",
            Self::AllowWithRedaction  => "ALLOW_WITH_REDACTION",
            Self::AllowWithLimits     => "ALLOW_WITH_LIMITS",
            Self::RequireApproval     => "REQUIRE_APPROVAL",
            Self::AllowWithAuditOnly  => "ALLOW_WITH_AUDIT_ONLY",
            Self::Quarantine          => "QUARANTINE",
        };
        f.write_str(s)
    }
}

/// The result of a policy evaluation request.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PolicyDecision {
    /// Whether the AI request is permitted to proceed.
    pub allowed: bool,

    /// The specific decision type from the policy engine.
    pub decision_type: DecisionType,

    /// Optional human-readable explanation of the decision.
    pub reason: Option<String>,

    /// When `decision_type` is [`DecisionType::AllowWithRedaction`], this
    /// field contains the prompt after PII and secrets have been removed.
    pub redacted_prompt: Option<String>,
}

impl PolicyDecision {
    /// Returns `true` if the request is allowed to proceed.
    pub fn is_allowed(&self) -> bool {
        self.allowed
    }
}
